torchtune/training/checkpointing/_checkpointer.py

```diff
@@ -374,27 +387,125 @@ def __init__(
             Path.joinpath(self._checkpoint_dir, "config.json").read_text()
         )
 
-        # save config.json to output_dir
-        save_config(self._output_dir, self._config)
+        # save all files in input_dir, except model weights and mapping, to output_dir
+        # this is useful to preserve the tokenizer, configs, license, etc.
+        copy_files(
+            self._checkpoint_dir,
+            Path.joinpath(self._output_dir, BASE_MODEL_DIRNAME),
+            ignore_suffixes=SUFFIXES_TO_NOT_COPY,
+        )
 
-        # recipe_checkpoint contains the recipe state. This should be available if
-        # resume_from_checkpoint is True
-        self._recipe_checkpoint = None
-        if self._resume_from_checkpoint:
-            if recipe_checkpoint is None:
-                raise ValueError(
-                    "If resume_from_checkpoint is True, recipe_checkpoint file must be provided."
+        # repo_id is necessary for when saving an adapter config, so its compatible with HF.
+        # This json file is produced and saved in the download step.
+        # contents are {"repo_id": "some_model/some_model_version"}
+        repo_id_path = Path.joinpath(self._checkpoint_dir, REPO_ID_FNAME).with_suffix(
+            ".json"
+        )
+        self.repo_id = None
+        if repo_id_path.exists():
+            with open(repo_id_path, "r") as json_file:
+                data = json.load(json_file)
+                self.repo_id = data.get("repo_id")
+
+        # e.g.
+        # checkpoint_files:
+        #   filename_format: model-{}-of-{}.safetensors
+        #   max_filename: 00191
+        if not isinstance(checkpoint_files, List):
+            # TODO: this can be a function instead of a class
+            formatted_checkpoint_files = FormattedCheckpointFiles.from_dict(
+                checkpoint_files
+            )
+            checkpoint_files = formatted_checkpoint_files.build_checkpoint_filenames()
+
+        # resume from ckpt
+        self._recipe_checkpoint = get_recipe_checkpoint_path(
+            output_dir=self._output_dir,
+            recipe_checkpoint=recipe_checkpoint,
+            resume_from_checkpoint=self._resume_from_checkpoint,
+        )
+
+        self._adapter_checkpoint = get_adapter_checkpoint_path(
+            output_dir=self._output_dir,
+            adapter_checkpoint=adapter_checkpoint,
+            resume_from_checkpoint=self._resume_from_checkpoint,
+            pattern=r"^epoch_(\d+)",
+        )
+
+        # e.g.checkpoint_files = [model-00001-of-00002.safetensors, model-00002-of-00002.safetensors]
+        # Case 1: no resuming from ckpt. Just load from checkpoint_dir.
```

Comment:
Thanks for enumerating the different cases here, but this is fairly convoluted. Why can't this be offloaded to a standalone utility?

Reply:
are you saying that the code can stay as it is, but have it as an utility? Or would you refactor the code itself?

Reply:
I am saying why not just put all of this (i.e. L436-481) in a utility like `def infer_checkpoint_paths(self, ...)`? Then you can put all this case 1, case 2, etc into a nice clear docstring in a single place. We could think about refactoring too.. the fact that we are calling `_validate_hf_checkpoint_files` like ten different ways is not my favorite

Reply:
done